python/rmm/_lib/lib.pyx

```diff
@@ -0,0 +1,314 @@
+                break
+            else:
+                frame = frame.f_back
+        line_number = frame.f_lineno
+        del frame
+    else:
+        filename = None
+        line_number = 0
+
+    cdef const char* file = <const char*>NULL
+    if filename is not None:
+        filename = filename.encode()
+        file = filename
+    cdef unsigned int line = line_number
+
+    return caller_pair(file, line)
+
+
+# API Functions
+def rmm_initialize(allocation_mode, initial_pool_size, enable_logging):
+    """
+    Initializes the RMM library by calling the librmm functions via Cython
+    """
+    global opts
+    opts = <rmmOptions_t*>malloc(sizeof(rmmOptions_t))
+    opts.allocation_mode = <rmmAllocationMode_t>allocation_mode
+    opts.initial_pool_size = <size_t>initial_pool_size
+    opts.enable_logging = <bool>enable_logging
+
+    with nogil:
+        rmm_error = rmmInitialize(
+            <rmmOptions_t *>opts
+        )
+
+    check_error(rmm_error)
+
+    return 0
+
+
+def rmm_finalize():
+    """
+    Finalizes the RMM library by calling the librmm functions via Cython
+    """
+    global opts
+    free(opts)
+    opts = NULL
+
+    with nogil:
+        rmm_error = rmmFinalize()
+
+    check_error(rmm_error)
+
+    return 0
+
+
+def rmm_is_initialized():
+    """
+    Returns True if RMM has been initialized, false otherwise by calling the
+    librmm functions via Cython
+    """
+    global opts
+
+    with nogil:
+        result = rmmIsInitialized(
+            <rmmOptions_t *>opts
+        )
+
+    return result
+
+
+def rmm_csv_log():
+    """
+    Returns a CSV log of all events logged by RMM, if logging is enabled by
+    calling the librmm functions via Cython
+    """
+    cdef size_t logsize = rmmLogSize()
+    cdef char* buf = <char*>malloc(logsize)
+
+    with nogil:
+        rmm_error = rmmGetLog(
+            <char*>buf,
+            <size_t>logsize
+        )
+
+    check_error(rmm_error)
+
+    cdef bytes py_buf = buf
+    free(buf)
+    return py_buf.decode("utf-8")
+
+
+cdef uintptr_t c_alloc(
+    size_t size, cudaStream_t stream
+) except? <uintptr_t>NULL:
+    """
+    Allocates size bytes using the RMM memory manager by calling the librmm
+    functions via Cython
+    """
+    cdef caller_pair tmp_caller_pair = _get_caller()
+    cdef const char* file = tmp_caller_pair.first
+    cdef unsigned int line = tmp_caller_pair.second
+
+    cdef void* ptr
+    with nogil:
+        rmm_error = rmmAlloc(
+            <void **>&ptr,
+            <size_t>size,
+            <cudaStream_t>stream,
+            <const char*>file,
+            <unsigned int>line
+        )
+
+    check_error(rmm_error)
+
+    return <uintptr_t>ptr
+
+
+def rmm_alloc(size, stream):
+    """
+    Allocates size bytes using the RMM memory manager by calling the librmm
+    functions via Cython
+    """
+    cdef size_t c_size = size
+    cdef cudaStream_t c_stream = <cudaStream_t><size_t>stream
+
+    cdef uintptr_t c_addr = c_alloc(
```

Comment:
Can you explain what the extra `c_alloc` layer is?

Reply:
This is a Cython API which is also exposed in the `lib.pxd` file, which allows another Cython project to depend on the `rmm` Python package and use these functions from within Cython. I though about getting rid of the `c_*` layer and just having the RMM C/C++ API and the Python API via Cython, but it felt like there was a bit of an awkward gap for Cython.